### ISSUE TITLE
Add HttpContext to IFileProvider methods

### ISSUE THREAD

**Issue by budcribar** [dotnet/aspnetcore#24402]
### Is your feature request related to a problem? Please describe.

I would like to implement the IFileProvider interface to serve a version of a file based on information stored in HttpContext. However, there is no way to get the context. 

### Describe the solution you'd like

Change the IFileProvider interface to optionally include the HttpContext as follows:
public IDirectoryContents GetDirectoryContents(string subpath, HttpContext context = null)
public IFileInfo GetFileInfo(string subpath, HttpContext context = null)
public IChangeToken Watch(string filter, HttpContext context = null))

This would also allow me to restrict a user from accessing a file based on what type of account they have.

**Comment by javiercn** [comment]
@budcribar thanks for contacting us.

The `IFileProvider` and related functionality lives in https://github.com/dotnet/runtime

That said, I would suggest you reconsider an alternative implementation. Trying to combine functionality from HttpContext and IFileProvider into the class is mixing very disparate concepts and its likely going to have issues.

I would suggest that instead you flow the HttpContext from the request to wherever you need it and use it to alter the response after retrieving a file or before allowing access to it.

I think we have some samples on how to achieve some of these things @Tratcher is that the case?

**Comment by budcribar** [comment]
Thanks for the response. For more detail I am trying to do the following:
app.UseStaticFiles(new StaticFileOptions {
       FileProvider = new MyCustomFileResolver()
 })

and I need to get the HttpContext into MyCustomFileResolver somehow in order to know which file to serve. Any ideas?

**Comment by Tratcher** [comment]
Yeah, StaticFiles is not designed for custom matching. Alternatives:
A) You can use an MVC controller and a FileResult
B) You can write a middleware that does your custom matching and either serves the file directly, or rewrites the path so that StaticFiles can serve it. See the DefaultFilesMiddleware for a rewrite example.
https://github.com/dotnet/aspnetcore/blob/5e49fc336e4464711ff2a8db2e00b9bafb518c78/src/Middleware/StaticFiles/src/DefaultFilesMiddleware.cs#L64

**Comment by budcribar** [comment]
I started with StaticFileMiddleware and got it working (see attached) but it took 15 files to get HttpContext into FileResolver.cs  (see attached)
[src.zip](https://github.com/dotnet/aspnetcore/files/5008329/src.zip)

It sure would be nice if there were a simpler solution. Any suggestions?

thanks,
bud

**Comment by Tratcher** [comment]
That seems overly complex. Can you summarize what you're actually trying to accomplish here? It seems like the only field you need off the request is a cookie? And you're using that split out multiple copies of blazor files?

**Comment by Pilchie** [comment]
Closing as we haven't heard back here.
